ls1intum/Artemis
# `Development`: Make command execution in local continuous integration compatible with Podman

### Checklist
#### General


#### Server

### Motivation and Context
Podman (https://podman.io/) is a drop-in replacement for Docker that allows to run containers without having to have `root` permissions (or having a `root`-privileged socket). This would require less privileges to run containers which provides improved security for running untrusted code in a production system. We should therefore aim to support both Docker and Podman if reasonably possible.

### Description
This PR fixes a small API incompatibility: Podman requires that at least one stream is attached to a command, or requires to explicitly set the command to be detached. The Docker API seems to implicitly set the detached flag internally.

Previous error message if both `attachStdout` and `attachStderr` are `false`:
```
[ream-1874955668] c.g.d.api.async.ResultCallbackTemplate   : Error during callback

com.github.dockerjava.api.exception.InternalServerErrorException: Status 500: {"cause":"invalid argument","message":"must provide at least one stream to attach to: invalid argument","response":500}

	at com.github.dockerjava.core.DefaultInvocationBuilder.execute(DefaultInvocationBuilder.java:247)
	at com.github.dockerjava.core.DefaultInvocationBuilder.lambda$executeAndStream$1(DefaultInvocationBuilder.java:269)
	at java.base/java.lang.Thread.run(Thread.java:840)
```


### Steps for Testing

Prerequisites:
- Setup with localci
- 1 Programming exercise

Since this is only specific to Podman, but test servers run Docker, you can only check that no functionality is broken:

1. Trigger a build for any submission or the solution/template in the exercise.
2. A build result should be received.

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage
unchanged
